x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -0,0 +1,363 @@
+      restoreState = {},
+    }: Partial<BackgroundSessionSavedObjectAttributes>,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    if (!name) throw new Error('Name is required');
+    if (!appId) throw new Error('AppId is required');
+    if (!urlGeneratorId) throw new Error('UrlGeneratorId is required');
+
+    this.logger.debug(`save | ${sessionId}`);
+
+    const attributes = {
+      name,
+      created,
+      expires,
+      status,
+      initialState,
+      restoreState,
+      idMapping: {},
+      urlGeneratorId,
+      appId,
+      sessionId,
+    };
+    const session = await savedObjectsClient.create<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      attributes,
+      { id: sessionId }
+    );
+
+    return session;
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public get = (sessionId: string, { savedObjectsClient }: BackgroundSessionDependencies) => {
+    this.logger.debug(`get | ${sessionId}`);
+    return savedObjectsClient.get<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      sessionId
+    );
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public find = (
+    options: BackgroundSessionFindOptions,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    return savedObjectsClient.find<BackgroundSessionSavedObjectAttributes>({
+      ...options,
+      type: BACKGROUND_SESSION_TYPE,
+    });
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public update = (
+    sessionId: string,
+    attributes: Partial<BackgroundSessionSavedObjectAttributes>,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    this.logger.debug(`update | ${sessionId}`);
+    return savedObjectsClient.update<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      sessionId,
+      attributes
+    );
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
+  public delete = (sessionId: string, { savedObjectsClient }: BackgroundSessionDependencies) => {
+    return savedObjectsClient.delete(BACKGROUND_SESSION_TYPE, sessionId);
+  };
+
+  /**
+   * Tracks the given search request/search ID in the saved session (if it exists). Otherwise, just
+   * store it in memory until a saved session exists.
+   * @internal
+   */
+  public trackId = async (
+    searchRequest: IKibanaSearchRequest,
+    searchId: string,
+    { sessionId, isStored }: ISearchOptions,
+    deps: BackgroundSessionDependencies
+  ) => {
+    if (!sessionId || !searchId) return;
+    this.logger.debug(`trackId | ${sessionId} | ${searchId}`);
+    const requestHash = createRequestHash(searchRequest.params);
+
+    // If there is already a saved object for this session, update it to include this request/ID.
+    // Otherwise, just update the in-memory mapping for this session for when the session is saved.
```

Comment:
**question** what happens if another instance of Kibana creates the saved session for this user? Will the in-memory mapping here get merged with the session the other instance created?

Reply:
Yes, that's the whole idea.
There can be multiple instances of Kibana, and as long as the requests and user match, the searchIds will be merged into this session object.

Reply:
Perfect!